### PR TITLE
Moved theme Map and added API to use it from other dependencies

### PR DESCRIPTION
**Pull request by JohanMabille**
This fixes some issues reported in #12812:

- matchBtackets
- smartIndent
- codeFolding
- foldGutter

This also adds a token for modes and themes registry.
This PR is not finished but a review would be appreciated to know if it goes in the right direction. If so I would add missing doc strings and redirection.

I also wonder if we should expose more methods from Mode via the IEditorModeRegistry interface.

### CODE REVIEW

**Review by fcollonval — CHANGES_REQUESTED** [review]
Using tokens is nice. But this requires changing a bit the current structure to not use static object. In particular the `editorThemeRegistry`. They should be propagated to the `EditorConfiguration`. Or the logic should be similar to the `EditorModeRegistry` that modifies a static list.

Otherwise, providing a different theme registry overriding the token will not have the expected outcome.

If we want to align more with other APIs, the generic pattern are:

- tokens that define registries
- registries provide factories
- factories instantiate objects

But that would require more work for the code editor.

**Comment by JohanMabille** [comment]
@fcollonval I removed the token API for now, so we don't have backward incompatible changes when we add it later.

> Or the logic should be similar to the EditorModeRegistry that modifies a static list.

Not sure to understand why this would work. The `EditorConfiguration` works with function that modifies the static list, so even if another extension defines another EditorModeRegistry, this later won't be used by the `EditorConfiguration` object and we'll run into the same issue as with the previous definition ot editorThemeRegistry.

My understanding is that we would also need to propagate the mode registry object to the editor configuration object to get the expected behavior. But that can be done after 4.0.

File: packages/codemirror/src/editortheme.ts [hunk 1]
```diff
@@ -76,3 +79,23 @@ export const jupyterTheme: Extension = [
   jupyterEditorTheme,
   syntaxHighlighting(jupyterHighlightStyle)
 ];
+
+export namespace Theme {
```
Comment (jtpio) [inline packages/codemirror/src/editortheme.ts:83]: It would be nice to add some basic docstrings here for consistency.

**Comment by fcollonval** [comment]
I'm taking over this PR.

I added a test for adding a mode (not yet pushed)
I was testing adding a theme but actually the reconfiguration of the editors are not happening. So I'm working to fix that.

**Comment by fcollonval** [comment]
bot please update snapshots

**Comment by fcollonval** [comment]
bot please update snapshots

**Comment by fcollonval** [comment]
bot please update snapshots

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Comment by fcollonval** [comment]
Kicking the CI

**Review by fcollonval — APPROVED** [review]

**Pull request merged**